File: src/my_py_pkg/my_py_pkg/battery.py
#!/usr/bin/env python3
import rclpy
from rclpy.node import Node
from my_robot_interfaces.srv import SetLed
from functools import partial

class BatteryNode(Node):
   def __init__(self):
      super().__init__("battery")

      self.battery_last_state_change = self.getTime()
      self.battery_empty = True
      self.callSetLedServer(3,True)
      timer_ = self.create_timer(0.1, self.simulateBattery)
      

      self.get_logger().info("battery has been started")


   def getTime(self):
      sec, nsec = self.get_clock().now().seconds_nanoseconds()
      
      return sec + nsec /1000000000.0
      
   def simulateBattery(self):
      now = self.getTime()
      if self.battery_empty:
         if now - self.battery_last_state_change > 6.0:
            self.battery_last_state_change = now
            self.battery_empty = False
            self.callSetLedServer(3, False)
            self.get_logger().info("Battery full")
      else:
         if now -self.battery_last_state_change > 4.0:
            self.battery_last_state_change = now
            self.battery_empty = True
            self.callSetLedServer(3, True)
            self.get_logger().info("Battery empty")
      
      
   def callSetLedServer(self, led:int, state:bool):
      cli_ = self.create_client(SetLed, "set_led")
      
      while not cli_.wait_for_service(timeout_sec=1.0):
         self.get_logger().warn("waiting for server...")
        
      request = SetLed.Request()
      request.led_number = led
      request.state = state

      future = cli_.call_async(request=request)
      future.add_done_callback(partial(self.callbackSetLed, led=led, state=state))
        
   def callbackSetLed(self, future, led, state):
      try:
         response = future.result()
         
         if response.success:
            self.get_logger().info(f"Led {led} has been set to {state}")
            
         else:
            self.get_logger().warn("Service call was successful but request not satisfied")
         
      except Exception as e:
         self.get_logger().error(f"Servce call faild with exception {e}") 

      


def main(args=None):
   rclpy.init(args=args)
   node = BatteryNode()
   rclpy.spin(node)
   rclpy.shutdown()

if __name__ == "__main__":
   main()

File: src/my_cpp_pkg/src/battery.cpp
#include "rclcpp/rclcpp.hpp"
#include "my_robot_interfaces/srv/set_led.hpp"

class batteryNode : public rclcpp::Node
{
public:
    batteryNode() : Node("battery"), battery_empty(true)
    {
        this->setLed(3, true);
        battery_last_state_change = this->get_clock()->now().seconds();
        timer_ = this->create_wall_timer(std::chrono::milliseconds(100), std::bind(&batteryNode::simulateBattery, this));
        RCLCPP_INFO(this->get_logger(), "battery has been started");
    }

    void setLed(int led, bool state)
    {
        threads_.push_back(std::thread(std::bind(&batteryNode::callSetLedServer, this, led, state)));
    }

    void callSetLedServer(int led, bool state)
    {
        auto client_ = this->create_client<my_robot_interfaces::srv::SetLed>("set_led");
        while (!client_->wait_for_service(std::chrono::seconds(1)))
        {
            RCLCPP_WARN(this->get_logger(), "Waiting for server ....");
        }

        auto request = std::make_shared<my_robot_interfaces::srv::SetLed_Request>();
        request->led_number = led;
        request->state = state;

        auto future = client_->async_send_request(request);

        try
        {
            auto response = future.get();
            RCLCPP_INFO(this->get_logger(), "led %d has been set to %s", led, state ? "on" : "off");
        }
        catch (const std::exception &e)
        {
            RCLCPP_ERROR(this->get_logger(), "Servce call failed");
        }
    }

    void simulateBattery()
    {
        double now = this->get_clock()->now().seconds();
        if (battery_empty)
        {
            if (now - battery_last_state_change > 6.0)
            {
                battery_last_state_change = now;
                battery_empty = false;
                this->setLed(3, false);
                RCLCPP_INFO(this->get_logger(), "Battery full");
            }
        }
        else
        {
            if (now - battery_last_state_change > 4.0)
            {
                battery_last_state_change = now;
                battery_empty = true;
                this->setLed(3, true);
                RCLCPP_INFO(this->get_logger(), "Battery empty");
            }
        }
    }

private:
    double battery_last_state_change;
    bool battery_empty;
    rclcpp::TimerBase::SharedPtr timer_;
    std::vector<std::thread> threads_;
};

int main(int argc, char **argv)
{
    rclcpp::init(argc, argv);
    auto node = std::make_shared<batteryNode>();
    rclcpp::spin(node);
    rclcpp::shutdown();
    return 0;
}